File: scripts/test_solutions/572_subtree_of_another_tree.py
def isSubtree(root, subRoot):
    def same(p, q):
        if not p and not q:
            return True
        if not p or not q or p.val != q.val:
            return False
        return same(p.left, q.left) and same(p.right, q.right)

    if not root:
        return False
    if same(root, subRoot):
        return True
    return isSubtree(root.left, subRoot) or isSubtree(root.right, subRoot)

File: scripts/test_solutions/572_subtree_of_another_tree.js
var isSubtree = function(root, subRoot) {
    function same(p, q) {
        if (!p && !q) return true;
        if (!p || !q || p.val !== q.val) return false;
        return same(p.left, q.left) && same(p.right, q.right);
    }
    if (!root) return false;
    if (same(root, subRoot)) return true;
    return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
};
